src/mux_client_impl/mod.rs

```diff
@@ -0,0 +1,129 @@
+#![allow(
+    missing_docs,
+    missing_debug_implementations,
+    rustdoc::broken_intra_doc_links,
+    unreachable_pub
+)]
+
+use std::borrow::Cow;
+use std::ffi::OsStr;
+use std::path;
+
+use shell_escape::escape;
+use tempfile::TempDir;
+
+use tokio::runtime;
+
+use openssh_mux_client::connection::Connection;
+pub use openssh_mux_client::connection::{ForwardType, Socket};
+
+use super::Error;
+
+pub(crate) mod builder;
+
+use super::fd::*;
+
+mod stdio;
+pub use stdio::{ChildStderr, ChildStdin, ChildStdout};
+
+use super::Stdio;
+
+mod command;
+pub use command::Command;
+
+mod child;
+pub use child::RemoteChild;
+
+#[derive(Debug)]
+pub struct Session {
+    /// TempDir will automatically removes the temporary dir on drop
+    tempdir: Option<TempDir>,
+}
+
+// TODO: UserKnownHostsFile for custom known host fingerprint.
+// TODO: Extract process output in Session::check(), Session::connect(), and Session::terminate().
```

Comment:
This can presumably also be shared with the process impl since it also waits on the ssh master process. I wonder if maybe it'd be worthwhile to extract all the logic around managing the master process into its own module — what do you think?

Reply:
Hmmm, I don't think so.

IMHO the implementation of `mux_client_impl::Session` is so different from `process_impl::Session` that it cannot be extracted out.